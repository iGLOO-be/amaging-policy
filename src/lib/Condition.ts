
import { ValidatorFn, ValidatorName } from "./Validators";

export
class Condition {
  public key: string;
  private validatorName: ValidatorName;
  private validator: ValidatorFn;
  private validatorArgs: any;

  constructor(key: string, validatorName: ValidatorName, validator: ValidatorFn, validatorArgs: any[]) {
    this.key = key;
    this.validatorName = validatorName;
    this.validator = validator;
    this.validatorArgs = validatorArgs;
  }

  public valid(value: any) {
    return this.validator(value, ...this.validatorArgs);
  }
}
