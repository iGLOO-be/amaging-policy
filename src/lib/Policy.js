
import isObject from 'lodash/isObject'
import filter from 'lodash/filter'
import extend from 'lodash/extend'
import toArray from 'lodash/toArray'
import Store from 'dottystore'
import validators from './validators'

/*
  Policy sample:

  { "expiration": "2007-12-01T12:00:00.000Z",
    "conditions": [
      ["starts-with", "$key", "user/eric/"],
      ["range-length", "$key", 10, 50],
      {"success_action_redirect": "http://johnsmith.s3.amazonaws.com/new_post.html"},
    ]
  }

*/

class PolicyError extends Error {
  constructor (type, data) {
    super(...arguments)

    this.type = type
    this.data = data
    this.name = 'PolicyError'
    this.message = (() => {
      switch (this.type) {
        case 'INVALID_KEY':
          return `Invalid value for key: ${this.data.key}`
        case 'INVALID_VALIDATOR_NAME':
          return `Invalid policy validator name: ${this.data.validator}`
      }
    })()
  }
}

class Condition {
  constructor (key, validatorName, validator, validatorArgs) {
    this.key = key
    this.validatorName = validatorName
    this.validator = validator
    this.validatorArgs = validatorArgs
  }
  valid (value) { return this.validator.apply(null, [value].concat(toArray(this.validatorArgs))) }
}

class Policy extends Store {
  static initClass () {
    this.validators = {}
  }
  static registerValidators (validators) {
    return (() => {
      const result = []
      for (let validatorName in validators) {
        result.push(Policy.registerValidator(validatorName, validators[validatorName]))
      }
      return result
    })()
  }
  static registerValidator (validatorName, validator) {
    Policy.validators[validatorName] = validator
  }
  static getValidator (validatorName) {
    return Policy.validators[validatorName]
  }

  constructor (policy) {
    super(...arguments)

    this._parsePolicy(policy)
  }

  set (key, value) {
    const conditions = this.getConditionForKey(key)

    for (let cond of Array.from(conditions || [])) {
      if (!cond.valid(value)) {
        throw new PolicyError('INVALID_KEY', {key})
      }
    }

    return super.set(...arguments)
  }

  _parsePolicy (policy = {}) {
    let key
    const data = {}
    const conditions = []

    for (let pol of Array.from(policy.conditions || [])) {
      if (Array.isArray(pol)) {
        const validatorName = pol[0]
        key = pol[1].replace(/^$/, '')
        const validatorArgs = pol.slice(2)

        const validator = Policy.getValidator(validatorName)
        if (!validator) {
          throw new PolicyError('INVALID_VALIDATOR_NAME', {validator})
        }

        const condition = new Condition(key, validatorName, validator, validatorArgs)
        conditions.push(condition)
      } else if (isObject(pol)) {
        extend(data, pol)
      }
    }

    this.conditions = conditions

    return (() => {
      const result = []
      for (key in data) {
        result.push(this.set(key, data[key]))
      }
      return result
    })()
  }

  getConditionForKey (key) {
    return filter(this.conditions, {key})
  }
}
Policy.initClass()

Policy.registerValidators(validators)

export default Policy
