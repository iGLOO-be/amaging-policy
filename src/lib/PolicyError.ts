
export interface IPolicyErrorData {
  key?: string;
  validator?: string;
}

export class PolicyError extends Error {
  public data: IPolicyErrorData | undefined;
  public type: string;

  constructor(type: string, data?: IPolicyErrorData) {
    super();

    this.type = type;
    this.data = data;
    this.name = "PolicyError";
    switch (this.type) {
      case "INVALID_KEY":
        this.message = `Invalid value for key: ${(data && data.key) || ""}`;
        break;
      case "INVALID_VALIDATOR_NAME":
        this.message = `Invalid policy validator name: ${(data && data.validator) || ""}`;
        break;
    }
  }
}
