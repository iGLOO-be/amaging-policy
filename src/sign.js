
import PolicyRepresentation from './lib/PolicyRepresentation'

export default function sign (accessKey, secret) {
  return new PolicyRepresentation(accessKey, secret)
}
