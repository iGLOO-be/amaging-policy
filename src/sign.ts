
import { PolicyRepresentation } from "./lib/PolicyRepresentation";

export default function sign(accessKey: string, secret: string) {
  return new PolicyRepresentation(accessKey, secret);
}
