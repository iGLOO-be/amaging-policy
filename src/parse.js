
import jwt from 'jsonwebtoken'
import Policy from './lib/Policy'
import debugFactory from 'debug'
import { promisify } from 'util'

const debug = debugFactory('amaging-policy:parse')
const jwtVerify = promisify(jwt.verify.bind(jwt))

export default async function parse (secret, token) {
  debug('Verify policy from JWT', token)
  let decoded

  try {
    decoded = await jwtVerify(token, secret)
  } catch (err) {
    debug('Got error during verify policy from JWT', err)

    if (err.name === 'TokenExpiredError' || err.name === 'JsonWebTokenError') {
      return false
    }

    throw err
  }

  debug('JWT valid with:', decoded)

  return new Policy(decoded.data)
}

export function getAccessKey (token) {
  let decoded

  try {
    decoded = jwt.decode(token)
  } catch (err) {
    if (err.name === 'TokenExpiredError' || err.name === 'JsonWebTokenError') {
      return
    }
    throw err
  }

  return (decoded && decoded.accessKey) || undefined
}
