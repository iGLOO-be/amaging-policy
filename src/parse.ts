
import debugFactory from "debug";
import { promisify } from "es6-promisify";
import jwt from "jsonwebtoken";
import { Policy } from "./lib/Policy";

const debug = debugFactory("amaging-policy:parse");
const jwtVerify: (token: string, secret: string) => Promise<IPolicyJWTCOntent> = promisify(jwt.verify.bind(jwt));

interface IPolicyJWTCOntent {
  accessKey?: string;
  data?: object;
}

export default async function parse(secret: string, token: string) {
  debug("Verify policy from JWT", token);
  let decoded;

  try {
    decoded = await jwtVerify(token, secret);
  } catch (err) {
    debug("Got error during verify policy from JWT", err);

    if (err.name === "TokenExpiredError" || err.name === "JsonWebTokenError") {
      return false;
    }

    throw err;
  }

  debug("JWT valid with:", decoded);

  return new Policy({
    conditions: Array.isArray(decoded.data) ? decoded.data : [],
  });
}

export function getAccessKey(token: string) {
  let decoded;

  try {
    decoded = jwt.decode(token) as IPolicyJWTCOntent;
  } catch (err) {
    if (err.name === "TokenExpiredError" || err.name === "JsonWebTokenError") {
      return;
    }
    throw err;
  }

  return (decoded && decoded.accessKey) || undefined;
}
