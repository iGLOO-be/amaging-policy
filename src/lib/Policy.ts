
import { Store } from "dottystore";
import extend from "lodash/extend";
import isObject from "lodash/isObject";
import { Condition } from "./Condition";
import { PolicyError } from "./PolicyError";
import { ValidatorName, validators } from "./validators";

/*
  Policy sample:s

  { "expiration": "2007-12-01T12:00:00.000Z",
    "conditions": [
      ["starts-with", "$key", "user/eric/"],
      ["range-length", "$key", 10, 50],
      {"success_action_redirect": "http://johnsmith.s3.amazonaws.com/new_post.html"},
    ]
  }

*/

export interface IPolicyObject {
  conditions?: IPolicyCondition[];
}
export type IPolicyCondition = [ValidatorName, string, ...any[]] | IPolicyData;
interface IPolicyData {
  [key: string]: string;
}

export
class Policy extends Store {
  public conditions: Condition[];

  constructor(policy?: IPolicyObject) {
    super();

    this.conditions = [];
    this._parsePolicy(policy);
  }

  public set(key: string, value: any) {
    const conditions = this.getConditionsForKey(key);

    for (const cond of Array.from(conditions || [])) {
      if (!cond.valid(value)) {
        throw new PolicyError("INVALID_KEY", {key});
      }
    }

    return super.set(key, value);
  }

  public _parsePolicy(policy: IPolicyObject = {}) {
    let key;
    const data: IPolicyData = {};
    const conditions = [];

    for (const pol of Array.from(policy.conditions || [])) {
      if (Array.isArray(pol)) {
        const validatorName = pol[0];
        key = pol[1].replace(/^$/, "");
        const validatorArgs = pol.slice(2);

        const validator = validators[validatorName];
        if (!validator) {
          throw new PolicyError("INVALID_VALIDATOR_NAME", {validator});
        }

        const condition = new Condition(key, validatorName, validator, validatorArgs);
        conditions.push(condition);
      } else if (isObject(pol)) {
        extend(data, pol);
      }
    }

    this.conditions = conditions;

    Object.keys(data).forEach((dataKey) => {
      this.set(dataKey, data[dataKey]);
    });
  }

  public getConditionsForKey(key: string): Condition[] {
    return this.conditions.filter((cond) => cond.key === key);
  }
}
