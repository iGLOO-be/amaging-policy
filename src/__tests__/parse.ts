
import parse, { getAccessKey } from "../parse";
import sign from "../sign";

import { Policy } from "../lib/Policy";

describe("parse", () => {
  it("Should correctly parse a policy", async () => {
    const jwt = await sign("key", "secret")
      .cond("eq", "key", "test")
      .toJWT();

    const policy = await parse("secret", jwt);
    if (!policy) {
      throw new Error("Unable to parse policy");
    }

    expect(policy).toBeInstanceOf(Policy);

    expect(policy.conditions).toMatchSnapshot();

    policy.set("key", "test");

    expect(() => {
      policy.set("key", "bar");
    }).toThrow("Invalid value for key: key");
  });
  it("Should results false if invalid token", async () => {
    expect(await parse("secret", "boom")).toEqual(false);
  });
  it("Should results false if expired token", async () => {
    const jwt = await sign("key", "secret").expiresIn(-1).toJWT();
    expect(await parse("secret", jwt)).toEqual(false);
  });
  it("Should results false if invalid secret", async () => {
    const jwt = await sign("key", "secret").toJWT();
    expect(await parse("bar", jwt)).toEqual(false);
  });
});

describe("getAccessKey", () => {
  it("Should get access key from JWT", async () => {
    const jwt = await sign("key", "secret").toJWT();
    const accesKey = getAccessKey(jwt);

    expect(accesKey).toEqual("key");
  });
});
