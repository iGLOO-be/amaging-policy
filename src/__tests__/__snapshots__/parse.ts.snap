// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse Should correctly parse a policy 1`] = `
Array [
  Condition {
    "key": "key",
    "validator": [Function],
    "validatorArgs": Array [
      "test",
    ],
    "validatorName": "eq",
  },
]
`;
